### DISCUSSION TITLE
AttributeError: module 'numpy' has no attribute 'int'

### DISCUSSION THREAD

**Discussion by lokiasgard07** [Marcello-Sega/pytim#405] [General]
Hi,
I was trying to find out the true interface of the liquid/vapor interface in my system using Pytim modules, but I am getting an error like below

import MDAnalysis as mda
import pytim
u = mda.Universe("md_0_1.gro")
interface = pytim.ITIM(u)
interface.writepdb('layers.pdb',centered='middle')
upper_layer = interface.layers[0,0]

AttributeError Traceback (most recent call last)
/tmp/ipykernel_80284/2357494757.py in
4 # initialize the interface, this has to be called only once
5 # before iterating over the trajectory!
----> 6 interface = pytim.ITIM(u)
7
8 # write a pdb with the information on the surface layers

~/anaconda3/lib/python3.9/site-packages/pytim/itim.py in init(self, universe, group, alpha, normal, molecular, max_layers, radii_dict, cluster_cut, cluster_threshold_density, extra_cluster_groups, info, centered, warnings, mesh, autoassign, **kargs)
217 cluster_threshold_density, extra_cluster_groups)
218 sanity.assign_normal(normal)
--> 219 sanity.assign_radii(radii_dict=radii_dict)
220
221 self.grid = None

~/anaconda3/lib/python3.9/site-packages/pytim/sanity_check.py in assign_radii(self, radii_dict)
146 self.patch_radii_dict()
147
--> 148 _missing_attributes(self.interface, self.interface.universe)
149
150 try:

~/anaconda3/lib/python3.9/site-packages/pytim/properties.py in _missing_attributes(interface, universe)
72 _check_missing_attribute(interface, 'elements', 'Elements', universe.atoms,
73 1)
---> 74 _extra_attributes(interface, universe)
75
76

~/anaconda3/lib/python3.9/site-packages/pytim/properties.py in _extra_attributes(interface, universe)
81 for key in attr.keys():
82 if key not in dir(universe.atoms):
---> 83 vals = np.zeros(len(universe.atoms), dtype=np.int) - 1
84 universe.add_TopologyAttr(attr[key](https://github.com/MDAnalysis/mdanalysis/discussions/vals))
85

~/anaconda3/lib/python3.9/site-packages/numpy/init.py in getattr(attr)
282 return Tester
283
--> 284 raise AttributeError("module {!r} has no attribute "
285 "{!r}".format(name, attr))
286

AttributeError: module 'numpy' has no attribute 'int'

Please tell me how I can fix it.
Thank you.

**Comment by Marcello-Sega** [comment]
Hi and thanks for reporting this! Could you let me know which numpy version
you are using? I think I know what the problem is and I’ll fix it as soon
as possible.

**Reply by lokiasgard07** [reply]
Thank you so much for the reply. I am using the version 1.26.4

**Reply by Marcello-Sega** [reply]
I guess you're using the conda package of pytim, right? The `np.int` issue should be already fixed in the master branch on github. I updated PyPI with the new 0.9.2 and I'm in the process of doing the same with conda-forge (although last time there were some problems, I hope they are fixed now).

Bottom line: you can (a) wait and reinstall from conda once I'm done (I'll keep you posted here), (b) install the latest using pip or (c) download and install from source from github.

**Reply by lokiasgard07** [reply]
Thank you so much. I have installed the new 0.9.2 version of Pytim, which works now.

**Reply by lokiasgard07** [reply]
Hi,
I am a little bit confused about which output file gives me the atom in the air/water interface to know the true interface. actually, I am not getting that atom that is in the interface.
Thank you

**Reply by Marcello-Sega** [reply]
There are two ways, see this script:

```python
import MDAnalysis as mda
import pytim
from pytim.datafiles import  WATER_GRO
import numpy as np
u = mda.Universe(WATER_GRO)
ox = u.select_atoms('name OW')
inter = pytim.GITIM(u, group=ox, alpha=2., molecular=True, max_layers=3, cluster_cut=3.5)
inter.writepdb('inter.pdb')
u.atoms.write('atoms.pdb')
```

If you open atoms.pdb, for example you'll see lines like:

```
ATOM     10  OW  SOL X   4      16.770  24.120  22.610  1.00  0.00      SYSTO   
ATOM     11  HW1 SOL X   4      16.990  24.360  23.560  1.00  0.00      SYSTH   
ATOM     12  HW2 SOL X   4      15.810  24.340  22.440  1.00  0.00      SYSTH   
ATOM     13  OW  SOL X   5      22.100  16.060  42.690  1.00  1.00      SYSTO   
ATOM     14  HW1 SOL X   5      22.440  16.980  42.460  1.00  1.00      SYSTH   
ATOM     15  HW2 SOL X   5      21.990  15.990  43.680  1.00  1.00      SYSTH   
ATOM     16  OW  SOL X   6       9.110  32.100   3.560  1.00  2.00      SYSTO   
ATOM     17  HW1 SOL X   6       8.200  32.150   3.140  1.00  2.00      SYSTH   
ATOM     18  HW2 SOL X   6       9.420  33.030   3.780  1.00  2.00      SYSTH   
ATOM     19  OW  SOL X   7      32.160  10.580   8.570  1.00  3.00      SYSTO   
ATOM     20  HW1 SOL X   7      31.690  10.770   9.430  1.00  3.00      SYSTH   
ATOM     21  HW2 SOL X   7      32.620  11.410   8.250  1.00  3.00      SYSTH   
[...]
ATOM   8623  OW  SOL X2875      49.670  25.640 149.980  1.00  0.50      SYSTO   
ATOM   8624  HW1 SOL X2875      48.880  25.400 149.410  1.00  0.50      SYSTH   
ATOM   8625  HW2 SOL X2875      49.340  25.960   0.870  1.00  0.50      SYSTH   
 
```

The last column reports the layer (0-> liquid, but not in any layer,  0.50 -> vapour phase,  1-> first layer, 2-> second layer, 3->third layer)

I hope this helps.

However, I might have spotted a problem with GITIM (for water) if you pass the whole universe instead of the group of oxygen atoms. ITIM is ok instead. I'll come back to you asap.

**Reply by Marcello-Sega** [reply]
Oh, and then you can use VMD to select those in different layers with the `beta` keyword:

![Screenshot 2024-05-14 at 11 17 38](https://github.com/Marcello-Sega/pytim/assets/466063/ddcac823-5328-4d82-993c-677b6e88f3ec)

to get something like this:

![Screenshot 2024-05-14 at 11 17 16](https://github.com/Marcello-Sega/pytim/assets/466063/75f938ee-3aeb-4965-9336-8d98c6fb6a96)

**Reply by Marcello-Sega** [reply]
So the "problem" with using all atoms (including hydrogens) for water is the usual one (I forget about it from time to time) that you have three co-planar, very close atoms in a water molecule, and this gives rise to pathologically large symplices, that's why I always use only the oxygen atoms with water and GITIM. I should find a way to exclude these atoms automatically, at some stage.

**Reply by lokiasgard07** [reply]
Thank you so much for your help and for giving me ideas about ITIM.

**Reply by lokiasgard07** [reply]
Hi,
After running the code you mentioned above, I got pdb, but some molecules are breaking.

![vmd](https://github.com/Marcello-Sega/pytim/assets/123500357/6e3b4712-e5f6-462f-a34a-2cd48f2e2505)

**Reply by Marcello-Sega** [reply]
That looks ok to me, you are working with periodic boundary conditions. If you prefer, for the sake of visualisation, you can center the liquid slabe before saving, have a look at the online manual:

https://marcello-sega.github.io/pytim/center.html

**Reply by lokiasgard07** [reply]
Hi,
Could you tell me how you decide this value (alpha=2, molecular=True, max_layers=3, cluster_cut=3.5)? I have tried using one system with cluster_cut=3.5. I didn't get any 0.50 value for the vapor phase, but when I change the value of cluster_cut to 2.5, I'm getting all the values in 0.5, however, in the case of 3, I get a smaller number of 0.50 values, so I don't know how to decide this thing.

What is the logic behind deciding the sphere, and how is it implemented in the code?

**Reply by Marcello-Sega** [reply]
For water, the first mininum of the rdf is located at around 3.5 Angstrom over a wide range of temperatures. This is usually a good choice to distinguish vapour from liquid molecules, at least at low temperatures, because most contact neighbors in a liquid cluster will be found below that distance. Of course, if you are simulating another liquid, you'll have to compute the rdf and check where the first minimum is located.

I wouldn't choose any other value than the location of the rdf minimum, unless you need to go to high temperatures. Closer to the critical point you should use more refined, density-based clustering methods (see https://pubs.rsc.org/en/content/articlehtml/2017/cp/c7cp02918g) ; In particular, for its application to the liquid/vapour interface of water see  https://pubs.rsc.org/en/content/articlehtml/2023/sm/d3sm00176h 
In pytim you can switch this on by passing, e.g., `cluster_cut=12. ,  cluster_threshold_density='auto'` , where `cluster_cut` now needs to be large enough to sample the local density. For water, usually 12 Angstrom is a good choice; don't go below 10 and much above 15. See the convergence plots in the 2017 paper if you're interested in reproducing the same for your system.

Regading your surprise at not finidng molecules in the vapor phase: at low temperatures (300 K) is completely normal to find from zero to 4-5 molecules in the vapour phase for a simulation of ~ 1000 molecules. Just convert the vapor and liquid densities from the coexistence diagram into number of molecules in your simulation box and you'll see what I mean.
